src/core/factories.ts

```diff
@@ -1,36 +1,69 @@
 import {action} from "mobx"
-import {extend} from "../utils"
+import {extend, fail} from "../utils"
 import {getNode, hasNode, NodeConstructor} from "./node"
 
 export type IModel = {
     $treenode: any // Actually Node, but that should not be exposed to the public...
 } & Object
 
-export interface IModelFactory<S, T> {
+export type IModelFactoryChecker = (snapshot: any) => boolean
+export type IModelFactoryDispatcher = (snapshot: any) => IModelFactory<any, any>
+export interface IModelFactoryConstructor<S, T>{
     (snapshot?: S, env?: Object): T & IModel
-    isModelFactory: true
+}
+
+export interface IModelFactory<S, T> extends IModelFactoryConstructor<S, T>{
     factoryName: string,
+    kind: string
+    is: IModelFactoryChecker
+    irreducible: boolean
+    dispatch: IModelFactoryDispatcher
     config: Object
+    isModelFactory: boolean
 }
 
-export function createFactory<S, T>(
+export function createFactoryConstructor<S, T>(
```

Comment:
I had to split this because we need to create 2 different types of factories. The "basic" ones which also attach administration objects, the other just attach the type system.